src/creature_instances.c

```diff
@@ -537,6 +548,133 @@ long process_creature_self_spell_casting(struct Thing* creatng)
     return 1;
 }
 
+/**
+ * @brief Check whether the given creature is suitable to cast ranged buff spell.
+ * This function is used for both combat and non-combat situations.
+ *
+ * @param creatng The creature being checked.
+ * @return CrInstance The instance index being set.
+ */
+CrInstance process_creature_ranged_buff_spell_casting(struct Thing* creatng)
+{
+    TRACE_THING(creatng);
+    SYNCDBG(8,"Processing %s(%d), act.st: %s, con.st: %s", thing_model_name(creatng), creatng->index,
+        creature_state_code_name(creatng->active_state), creature_state_code_name(creatng->continue_state));
+
+    if ((creatng->alloc_flags & TAlF_IsControlled) != 0)
+    {
+        // If this creature is under player's control.
+        return CrInst_NULL;
+    }
+
+    struct CreatureControl* cctrl = creature_control_get_from_thing(creatng);
+    if (cctrl->instance_id != CrInst_NULL) {
+        SYNCDBG(15, "This creature has an instance %s. Abort.", creature_instance_code_name(cctrl->instance_id));
+        return CrInst_NULL;
+    }
+
+    // Check validation of the caster.
+    if (creature_is_fleeing_combat(creatng) || creature_affected_by_spell(creatng, SplK_Chicken) ||
+        creature_is_being_unconscious(creatng) || creature_is_dying(creatng) ||
+        thing_is_picked_up(creatng) || creature_is_being_dropped(creatng) ||
+        creature_is_being_sacrificed(creatng) || creature_is_being_summoned(creatng) ||
+        // Creature who is leaving doesn't care about any allies.
+        creatng->continue_state == CrSt_CreatureLeaves ||
+        creatng->active_state == CrSt_CreatureLeavingDungeon ||
+        creatng->active_state == CrSt_CreatureScavengedDisappear )
+    {
+        return CrInst_NULL;
+    }
+
+    if (!creature_instance_is_available(creatng, CrInst_RANGED_HEAL) ||
+        !creature_instance_has_reset(creatng, CrInst_RANGED_HEAL))
+    {
+        return CrInst_NULL;
+    }
+
+    struct Thing* best_choice = NULL;
+    if (creature_would_benefit_from_healing(creatng))
+    {
+        // Save itself first.
+        best_choice = creatng;
+    }
+    else
+    {
+        // Search nearby area, collect candidates.
+        struct Dungeon* dungeon = get_players_num_dungeon(creatng->owner);
+        int creature_idx = dungeon->creatr_list_start;
+        int k = 0;
+        while (creature_idx != 0)
+        {
+            struct Thing* candidate_thing = thing_get(creature_idx);
+            struct CreatureControl* candidate_ctrl = creature_control_get_from_thing(candidate_thing);
+            creature_idx = candidate_ctrl->players_next_creature_idx;
+            if (creature_control_invalid(candidate_ctrl))
+            {
+                ERRORLOG("Invalid creature control");
+                break;
+            }
+            // Check validation of candidate's state.
+            if (candidate_thing->index == creatng->index ||
```

Comment:
This whole `if (`, with the exception of `creature_would_benefit_from_healing` would be well placed in a new function. Something like `creature_is_valid_buff_target`.

Reply:
I will re-design.